### PR TITLE
Swap from easyhid to cython-hidapi

### PR DESCRIPTION
**Pull request by nickswalker**
An alternative to #55 , also addressing macOS compatibility. This PR also includes a fix for opening by path under macOS/Windows, which don't use filesystem paths for devices.

@peter-mitrano-ar

## Summary by Sourcery

Switch the HID integration to cython-hidapi and improve cross-platform device discovery, opening, and setup guidance.

Bug Fixes:
- Fix HID device opening by path across Linux, macOS, and Windows by supporting platform-specific opaque device identifiers while retaining Linux path resolution.

Enhancements:
- Replace easyhid with the cython-hidapi bindings and defer backend loading with clear diagnostics for missing or conflicting HID packages.
- Prefer the Linux hidraw backend and apply nonblocking configuration when opening devices.
- Document bundled HID dependencies, platform-independent setup, device visibility diagnostics, and updated Linux permissions.
- Document opening devices with common axis conventions and customize axis mappings for application-specific coordinate systems.

Build:
- Update the project dependency from easyhid to hidapi and remove CI steps that install system hidapi libraries.

Documentation:
- Update the README, troubleshooting guides, and contributor documentation for the new hidapi backend, CLI diagnostics, cross-platform paths, and udev permissions.

### CODE REVIEW

File: pyspacemouse/_hid.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    if sys.platform.startswith("linux"):
+        try:
+            import hidraw
+
+            module = hidraw
+        except ImportError:
+            # Older cython-hidapi releases, and source builds configured with
+            # --with-libusb, only provide `hid`.
+            module = None
```
Comment (peter-mitrano-ar) [inline pyspacemouse/_hid.py:61]: BTW I tested commenting this out, and the library still works on linux but you need either more/different udev rules, or sudo, or something else otherwise you get an error. Hence I've added the path we tried opening with to the error so it says: 
```
Failed to open SpaceMouse: Failed to open device: b'3-7:1.0'
```

**Review by peter-mitrano-ar — APPROVED** [review]
This changes dependencies, so I am hesitant to merge just yet. I would like to let it sit, and maybe I will have time to test upgrading with some projects at work. I also think we should do a release first, bumping to 2.1.0, since we have not had a release since the 2.0.0 re-write which had some problems. That release would also introduce the concept of axis conventions and the warning about "LEGACY". Then we can merge this after. Sound good?

File: pyspacemouse/_hid.py (hunk 1 above)
Comment (nickswalker) [reply pyspacemouse/_hid.py:61]: Good idea.

Looking more at the cython wheels, they always provide `hidraw` (raw kernel hid backend, allows multiple processes to read the same device), so maybe catching the importerror is overly defensive.

Based on reading and testing, I think a good recommendation for the updated rules that cover libusb fallback would be:
```
SUBSYSTEM=="hidraw", ATTRS{idVendor}=="256f", MODE="0664", GROUP="input", TAG+="uaccess"
SUBSYSTEM=="usb", ENV{DEVTYPE}=="usb_device", ATTR{idVendor}=="256f", MODE="0664", GROUP="input", TAG+="uaccess"
```

`TAG+` is an ACL mechanism that grants any user with physical access can use the device. Any shell/process started under a GUI login will have access. The input group and permissions (0664 so only group members can write) cover users who want to start a process via a remote shell, which doesn't trigger the ACL.

**Comment by nickswalker** [comment]
Agreed on doing an interim release and letting this sit.

I updated the documented udev rules after looking into them. Users' old rules should keep working because hidraw is the default. The new ones just add an entry for libusb fallback, add ACL based access and tighten up the default breadth/permissions of the rule

**Review by peter-mitrano-ar — CHANGES_REQUESTED** [review]
Hm I'm not so sure about this latest commit. Existing users will already be in the plugdev group, and I don't see any benefit to switching to input. Also, adding the TAG uaccess will restrict things so it doesn't work via SSH, right? I don't see how "Use ACL access for easy/immediate interactive access" is true. If this is about increasing security, I am not really convinced it's a good idea.

**Comment by peter-mitrano-ar** [comment]
Ah. sorry. just saw your comment. I see that you suggest adding a new rule for "usb", but you also changed the hidraw one, which I'm still not sure about. And even for the "usb" rule, I would again not at the uaccess tag, it just adds a new edge case that will confuse people with "permission denied". But maybe I misunderstand, please let me know what you think.

**Comment by peter-mitrano-ar** [comment]
Wait sorry, maybe I just need to test it . read your _other_ comment and now I'm not sure...

**Comment by peter-mitrano-ar** [comment]
Ok so I did some testing... 

No udev rule -- permission denied, obviously.
Old udev rule -- works, and my user was already in plugdev because I've set that up before. By user is _not_ in input.
New udev rule -- DOES NOT WORK (makes sense)
New udev rule but with plugdev -- Works, including over SSH.

Now I decided to test the case where I _omit_ the MODE and GROUP entirely. Normally this would fail, and it still does. It seems the ACL stuff is not working, at least for me on this machine. Did you test it yourself? I think either way we should change back to plugdev to avoid breaking things. Sorry for the message spam.

**Comment by nickswalker** [comment]
>  I don't see how "Use ACL access for easy/immediate interactive access" is true.

`uaccess` is preferable to `GROUP` because it avoids the user having to add themselves to a group to have access. One set of rules would immediately cover all users on the machine (even those not created yet), provided they are physically logged into the machine. Based on my reading, this is the recommended format for udev rules to give device access nowadays: https://wiki.archlinux.org/title/Udev#Allowing_regular_users_to_use_devices

>  It seems the ACL stuff is not working, at least for me on this machine. Did you test it yourself?

I had my rules at priority 50, which turns out to be a necessary condition for the ACLs rules to process: https://github.com/systemd/systemd/issues/4288. I updated the priority used by the docs.

> Also, adding the TAG uaccess will restrict things so it doesn't work via SSH, right?

I tested this further and observed that the ACL is only removed when the user _logs out of their GUI desktop session_. Locking the screen has no impact. SSH-ing in under the same account while that GUI session is still alive will also happily allow the user access to the device. That must cover the vast majority of common use cases, so I went ahead and dropped `GROUP` from the default recommended rule.

> and I don't see any benefit to switching to input. 

You're right, `input` is wildly worse (I picked it because it was the recommendation in the existing troubleshooting docs). A user adding themselves to the group would be making it easy to sniff all their device input. Given that the ACLs should handle all but very odd uses (i.e. wanting to use the mouse while SSH-ed into a machine as a different user than the currently logged in one, or when no user is logged in at all), I think it makes sense to suggest that users create a device-specific group and add their user only if they find a need for it (and read the troubleshooting page).

>  either way we should change back to plugdev to avoid breaking things

`plugdev` is an anachronism (see it being dropped from debian 8 years ago https://bugs-devel.debian.org/cgi-bin/bugreport.cgi?bug=897916;msg=10), so I think it makes sense to not use it to avoid confusion (it was only slightly more targeted than `input` in its day). 




Could you test out the plain ACL rules again at a lower priority and see if they work as expected?  In any case changing the recommendations won't have any impact on current users' rules, but I think it will remove a step that could go wrong for new users (creating a group, adding the users that need the device to it).

**Review by peter-mitrano-ar — APPROVED** [review]
Thanks for testing more and explaining -- the uaccess tag now works for me. Not having to add yourself to a group is great, although to be honest I wish we could do this without sudo entirely. I'm mostly happy to get rid of the extra system dependency! I re-ordered some sections of the README since most of the devs I know use linux, and the udev rule was pretty far down.
